llvm/test/Transforms/InstSimplify/cmp-alloca-offsets.ll

```diff
@@ -249,8 +250,10 @@ define i1 @zst_alloca_middle() {
 ; CHECK-LABEL: @zst_alloca_middle(
 ; CHECK-NEXT:    [[A:%.*]] = alloca i64, align 8
 ; CHECK-NEXT:    [[A2:%.*]] = alloca {}, align 8
+; CHECK-NEXT:    [[GEP:%.*]] = getelementptr i8, ptr [[A]], i64 4
+; CHECK-NEXT:    [[CMP:%.*]] = icmp eq ptr [[GEP]], [[A2]]
 ; CHECK-NEXT:    call void @escape(ptr [[A]], ptr [[A2]])
-; CHECK-NEXT:    ret i1 false
+; CHECK-NEXT:    ret i1 [[CMP]]
```

Comment:
Can you add an InstCombine test to demonstrate that this case can be folded using KnownBits (`??000` != `??100`)?

Reply:
Oops, this was actually a mistake in the test. The zero-size allocation was supposed to have `align 1`. It's weird that LLVM infers `align 8` for it by default. I adjusted the test.

InstCombine actually fails to handle this (see https://llvm.godbolt.org/z/WMPKdnrMd for variant without allocas) due to a weakness in foldICmpUsingKnownBits. I'll work on a fix.

Reply:
Here's the fix: https://github.com/llvm/llvm-project/pull/115874